README.md
# yafwi [<img src="https://img.shields.io/gitlab/pipeline/alen/yafwi/main?gitlab_url=https%3A%2F%2Fgitlab.home.alen.sh%2F&label=Gitlab%20CI&style=flat-square" align="right">](https://gitlab.home.alen.sh/alen/yafwi) [<img src="https://img.shields.io/travis/buhanec/yafwi/main.svg?label=Travis+CI&style=flat-square" align="right">](https://travis-ci.org/buhanec/yafwi) [<img src="https://img.shields.io/azure-devops/build/buhanec/5aa80764-f0fc-4e76-8e87-bae1eeba6574/3/main?label=Azure%20DevOps&style=flat-square" align="right">](https://dev.azure.com/buhanec/yafwi/_build)

## Yet Another Fixed Width Integer (Library)

Simple fixed with integers for developer experimentation.

# Usage

## Installation

Available through [yafwi - PyPI](https://pypi.org/project/yafwi/) using `pip install yafwi`.

## Some Examples

```python
>>> from yafwi import *
>>> from yafwi import __all__
>>> __all__
('FixedWidthInt', 'BaseFixedWidthInt', 'generate_int', 
 'int8', 'int16', 'int32', 'int64', 'int128', 'int256', 
 'uint8', 'uint16', 'uint32', 'uint64', 'uint128', 'uint256', 
 'sbyte', 'byte', 'short', 'ushort', 'int_', 'uint', 'long', 'ulong')

>>> int8(120)
int8(120)

>>> int8(120) + 120  # Arithmetic with Python int preserves type
int8(-16)
>>> int8(120) + int16(120)  # Arithmetic will take result in larger width
int16(240)

>>> int16.max, int16.min  # Utility sentinels
(int16(32767), int16(-32768))
>>> uint32.min, uint32.max
(uint32(0), uint32(4294967295))

>>> int8(0b11110000)  # Utility represenations
int8(-16)
>>> int8(0b11110000).bin
'0b11110000'
>>> int8(0b11110000).hex
'0xf0'

>>> (~int8(0b10010110)).bin
'0b01101001'
>>> (int8(0b10010110) >> 2).bin
'0b11100101'
>>> (int8(0b10010110) << 2).bin
'0b01011000'
>>> ((uint32(uint8.max) << 10 | uint8.max) ^ uint32.max - (1 << 31)).bin
'0b01111111111111000000001100000000'

>>> bytes(uint32(8))  # Output depends on system byteorder
b'\x08\x00\x00\x00'  
>>> uint32(8).hex  # Output independent of system byteorder
'0x00000008'
```

## Convenient Aliases

```python
>>> from yafwi import *

>>> uint is uint32
True
>>> long is int64
True
>>> ushort is uint16
True
[...]
```

## Arbitrary Sizes

```python
>>> import yafwi

>>> yafwi.generate_int(1024, unsigned=True)
<class 'yafwi.uint1024'>
>>> yafwi.int512
<class 'yafwi.int512'>

>>> yafwi.int3.max, yafwi.int3.min  # Including some fun ones... 
(int3(3), int3(-4))
```

# Platform Independence

If run on a big-endian system, the bytes associated with the numbers will be different, but higher-level functionality should remain unchanged.


CHANGELOG.md
# Changelog

## v1.0.1 (2020-03-10)
- Clean up typing, README typos

## v1.0.0 (2020-07-16)
- Initial release


test_yafwi.py
from hypothesis import given
from hypothesis.strategies import integers, sampled_from
import pytest

from yafwi import *


@pytest.mark.parametrize('cls, expected', [
    (sbyte, int8),
    (byte, uint8),
    (short, int16),
    (ushort, uint16),
    (int_, int32),
    (uint, uint32),
    (long, int64),
    (ulong, uint64),
])
def test_aliases(cls: FixedWidthInt, expected: FixedWidthInt):
    assert cls is expected


@pytest.mark.parametrize('val, other, expected', [
    (int8(-128), -128, True),
    (int8(127), 127, True),
])
def test_equality(val: FixedWidthInt, other: int, expected: bool):
    assert (val == other) is expected


@pytest.mark.parametrize('cls, min_val, max_val', [
    (int8, -128, 127),
    (uint8, 0, 255),
    (int16, -32768, 32767),
    (uint16, 0, 65535),
    (int32, -2147483648, 2147483647),
    (uint32, 0, 4294967295),
    (int64, -9223372036854775808, 9223372036854775807),
    (uint64, 0, 18446744073709551615),
    (int128, -170141183460469231731687303715884105728, 170141183460469231731687303715884105727),
    (uint128, 0, 340282366920938463463374607431768211455),
    (int256, -57896044618658097711785492504343953926634992332820282019728792003956564819968, 57896044618658097711785492504343953926634992332820282019728792003956564819967),
    (uint256, 0, 115792089237316195423570985008687907853269984665640564039457584007913129639935),
])
def test_limits(cls: FixedWidthInt, min_val: int, max_val: int):
    assert isinstance(cls.min, cls)
    assert cls.min == min_val
    assert isinstance(cls.max, cls)
    assert cls.max == max_val


@pytest.mark.parametrize('cls', [
    int8, uint8, int16, uint16, int32, uint32,
    int64, uint64, int128, uint128, int256, uint256,
])
def test_overflow(cls: FixedWidthInt):
    assert cls.max + 1 == cls.min
    assert cls.min - 1 == cls.max


@given(reference=sampled_from([int8, uint8, int16, uint16,
                               int32, uint32, int64, uint64]),
       value=integers(min_value=int(int64.min) * 100,
                      max_value=int(uint64.max) * 100))
def test_generated(reference, value):
    generated = generate_int(reference.width, reference.unsigned)
    assert generated(value) == reference(value)


tests/test_yafwi.py
from hypothesis import assume, given
from hypothesis.strategies import composite, integers, sampled_from
import pytest

from yafwi import *


@pytest.mark.parametrize('cls, expected', [
    (sbyte, int8),
    (byte, uint8),
    (short, int16),
    (ushort, uint16),
    (int_, int32),
    (uint, uint32),
    (long, int64),
    (ulong, uint64),
])
def test_aliases(cls: FixedWidthInt, expected: FixedWidthInt):
    assert cls is expected


@pytest.mark.parametrize('cls, min_val, max_val', [
    (int8, -128, 127),
    (uint8, 0, 255),
    (int16, -32768, 32767),
    (uint16, 0, 65535),
    (int32, -2147483648, 2147483647),
    (uint32, 0, 4294967295),
    (int64, -9223372036854775808, 9223372036854775807),
    (uint64, 0, 18446744073709551615),
    (int128, -170141183460469231731687303715884105728, 170141183460469231731687303715884105727),
    (uint128, 0, 340282366920938463463374607431768211455),
    (int256, -57896044618658097711785492504343953926634992332820282019728792003956564819968, 57896044618658097711785492504343953926634992332820282019728792003956564819967),
    (uint256, 0, 115792089237316195423570985008687907853269984665640564039457584007913129639935),
])
def test_limits(cls: FixedWidthInt, min_val: int, max_val: int):
    assert isinstance(cls.min, cls)
    assert cls.min == min_val
    assert isinstance(cls.max, cls)
    assert cls.max == max_val


@pytest.mark.parametrize('cls', [
    int8, uint8, int16, uint16, int32, uint32,
    int64, uint64, int128, uint128, int256, uint256,
])
def test_overflow(cls: FixedWidthInt):
    assert cls.max + 1 == cls.min
    assert cls.min - 1 == cls.max


@given(reference=sampled_from((int8, uint8, int16, uint16,
                               int32, uint32, int64, uint64)),
       value=integers(min_value=int(int64.min) * 100,
                      max_value=int(uint64.max) * 100))
def test_generated(reference, value):
    generated = generate_int(reference.width, reference.unsigned)
    assert generated(value) == reference(value)


@composite
def ref_and_value(draw, references=(int8, uint8, int16, uint16,
                                    int32, uint32, int64, uint64,
                                    int128, uint128, int256, uint256)):
    reference = draw(sampled_from(references))
    value = draw(integers(min_value=reference.min, max_value=reference.max))
    return reference, value


@given(ref=ref_and_value())
def test_eq(ref):
    reference, value = ref
    assert reference(value) == value
    assert reference(value) >= value
    assert reference(value) <= value
    assert reference(value) == reference(value)
    assert reference(value) >= reference(value)
    assert reference(value) <= reference(value)


@composite
def ref_and_neq_values(draw, references=(int8, uint8, int16, uint16,
                                         int32, uint32, int64, uint64,
                                         int128, uint128, int256, uint256)):
    reference = draw(sampled_from(references))
    value1 = draw(integers(min_value=reference.min, max_value=reference.max))
    value2 = draw(integers(min_value=reference.min, max_value=reference.max))
    assume(value1 != value2)
    return reference, value1, value2


@given(ref=ref_and_neq_values())
def test_neq(ref):
    reference, value1, value2 = ref
    assert reference(value1) != value2
    assert reference(value2) != value1
    assert reference(value1) != reference(value2)


@composite
def ref_and_lt_values(draw, references=(int8, uint8, int16, uint16,
                                         int32, uint32, int64, uint64,
                                         int128, uint128, int256, uint256)):
    reference = draw(sampled_from(references))
    value1 = draw(integers(min_value=reference.min, max_value=reference.max))
    value2 = draw(integers(min_value=reference.min, max_value=reference.max))
    assume(value1 < value2)
    return reference, value1, value2


@given(ref=ref_and_lt_values())
def test_lt_gt(ref):
    reference, value1, value2 = ref
    assert reference(value1) < value2
    assert reference(value1) <= value2
    assert reference(value2) > value1
    assert reference(value2) >= value1
    assert reference(value1) < reference(value2)
    assert reference(value1) <= reference(value2)
    assert reference(value2) > reference(value1)
    assert reference(value2) >= reference(value1)


yafwi.py
"""Yet another fixed with integer."""

__all__ = ('FixedWidthInt', 'BaseFixedWidthInt', 'generate_int',
           'int8', 'int16', 'int32', 'int64', 'int128', 'int256',
           'uint8', 'uint16', 'uint32', 'uint64', 'uint128', 'uint256',
           'sbyte', 'byte',
           'short', 'ushort',
           'int_', 'uint',
           'long', 'ulong')


from ctypes import (c_int16, c_int32, c_int64, c_int8,
                    c_uint16, c_uint32, c_uint64, c_uint8)
from functools import wraps
from sys import byteorder
from typing import Protocol, Tuple, Type, TypeVar

T = TypeVar('T')


class CTypeInt(Protocol):
    value: int

    def __init__(self, value: int) -> None:
        ...


class FixedWidthInt(type):
    _raw: CTypeInt
    _width: int
    _unsigned: bool

    @property
    def width(cls) -> int:
        return cls._width

    @property
    def unsigned(cls) -> bool:
        return cls._unsigned

    @property
    def max(cls) -> int:
        if cls.unsigned:
            return cls(2 ** cls._width - 1)
        return cls(2 ** (cls._width - 1) - 1)

    @property
    def min(cls) -> int:
        if cls.unsigned:
            return cls(0)
        return cls(-(2 ** (cls._width - 1)))


def take_wider(fn):
    @wraps(fn)
    def wrapper(self: 'BaseFixedWidthInt', other: int):
        if (isinstance(other, BaseFixedWidthInt)
                and (other.width > self.width
                     or other.width == self.width
                     and other.unsigned
                     and not self.unsigned)):
            return NotImplemented
        return fn(self, other)

    return wrapper


class BaseFixedWidthInt(int, metaclass=FixedWidthInt):

    def __new__(cls: Type[T], value: int) -> T:
        if cls is BaseFixedWidthInt:
            raise RuntimeError('Use concrete implementation, not base _Int')

        raw = cls._raw(value)
        if isinstance(raw, int):
            # noinspection PyArgumentList
            return int.__new__(cls, raw)
        # noinspection PyArgumentList
        return int.__new__(cls, raw.value)

    @property
    def width(self: T) -> T:
        # noinspection PyTypeChecker
        return type(self).width

    @property
    def unsigned(self: T) -> bool:
        # noinspection PyTypeChecker
        return type(self).unsigned

    @property
    def max(self: T) -> T:
        # noinspection PyTypeChecker
        return type(self).max

    @property
    def min(self: T) -> T:
        # noinspection PyTypeChecker
        return type(self).min

    # Conversion / reprs

    def __bytes__(self) -> bytes:
        return self.to_bytes(self.width // 8,
                             byteorder=byteorder,
                             signed=not self.unsigned)

    @property
    def bin(self) -> str:
        if byteorder == 'little':
            return '0b' + ''.join(f'{b:08b}' for b in reversed(bytes(self)))
        return '0b' + ''.join(f'{b:08b}' for b in bytes(self))

    @property
    def hex(self) -> str:
        if byteorder == 'little':
            return '0x' + ''.join(f'{b:02x}' for b in reversed(bytes(self)))
        return '0x' + ''.join(f'{b:02x}' for b in bytes(self))

    def __repr__(self: T) -> str:
        if self.unsigned:
            prefix = 'u'
        else:
            prefix = ''
        return f'{prefix}int{self.width}({super().__repr__()})'

    # Arithmetic

    @take_wider
    def __add__(self: T, other: int) -> T:
        return type(self)(super().__add__(other))

    @take_wider
    def __radd__(self: T, other: int) -> T:
        return type(self)(super().__radd__(other))

    @take_wider
    def __sub__(self: T, other: int) -> T:
        return type(self)(super().__sub__(other))

    @take_wider
    def __rsub__(self: T, other: int) -> T:
        return type(self)(super().__rsub__(other))

    @take_wider
    def __mul__(self: T, other: int) -> T:
        return type(self)(super().__mul__(other))

    @take_wider
    def __rmul__(self: T, other: int) -> T:
        return type(self)(super().__rmul__(other))

    def __divmod__(self: T, other: int) -> Tuple[T, T]:
        div, mod = super().__divmod__(other)
        return type(self)(div), type(self)(mod)

    def __floordiv__(self: T, other: int) -> T:
        return type(self)(super().__floordiv__(other))

    def __mod__(self: T, other: int) -> T:
        return type(self)(super().__mod__(other))

    def __pow__(self: T, power: int, modulo=None) -> T:
        return type(self)(super().__pow__(power, modulo))

    # Bitwise ops

    @take_wider
    def __and__(self: T, other: int) -> T:
        return type(self)(super().__and__(other))

    @take_wider
    def __rand__(self: T, other: int) -> T:
        return type(self)(super().__rand__(other))

    @take_wider
    def __or__(self: T, other: int) -> T:
        return type(self)(super().__or__(other))

    @take_wider
    def __ror__(self: T, other: int) -> T:
        return type(self)(super().__ror__(other))

    @take_wider
    def __xor__(self: T, other: int) -> T:
        return type(self)(super().__xor__(other))

    @take_wider
    def __rxor__(self: T, other: int) -> T:
        return type(self)(super().__rxor__(other))

    def __lshift__(self: T, other: int) -> T:
        return type(self)(super().__lshift__(other))

    def __rshift__(self: T, other: int) -> T:
        return type(self)(super().__rshift__(other))

    def __invert__(self: T) -> T:
        return type(self)(super().__invert__())

    # Other maths / conversions

    def __abs__(self: T) -> T:
        return type(self)(super().__abs__())

    def __ceil__(self: T) -> T:
        return type(self)(self)

    def __floor__(self: T) -> T:
        return type(self)(self)

    def __index__(self: T) -> T:
        return type(self)(self)

    def __trunc__(self: T) -> T:
        return type(self)(self)

    def __round__(self: T, n=None) -> T:
        return type(self)(super().__round__(n))

    def __pos__(self: T) -> T:
        return type(self)(super().__pos__())

    def __neg__(self: T) -> T:
        return type(self)(super().__neg__())


# noinspection PyPep8Naming
class int8(BaseFixedWidthInt):
    _raw = c_int8
    _width = 8
    _unsigned = False


# noinspection PyPep8Naming
class int16(BaseFixedWidthInt):
    _raw = c_int16
    _width = 16
    _unsigned = False


# noinspection PyPep8Naming
class int32(BaseFixedWidthInt):
    _raw = c_int32
    _width = 32
    _unsigned = False


# noinspection PyPep8Naming
class int64(BaseFixedWidthInt):
    _raw = c_int64
    _width = 64
    _unsigned = False


# noinspection PyPep8Naming
class uint8(BaseFixedWidthInt):
    _raw = c_uint8
    _width = 8
    _unsigned = True


# noinspection PyPep8Naming
class uint16(BaseFixedWidthInt):
    _raw = c_uint16
    _width = 16
    _unsigned = True


# noinspection PyPep8Naming
class uint32(BaseFixedWidthInt):
    _raw = c_uint32
    _width = 32
    _unsigned = True


# noinspection PyPep8Naming
class uint64(BaseFixedWidthInt):
    _raw = c_uint64
    _width = 64
    _unsigned = True


def generate_int(width: int, unsigned: bool) -> Type[BaseFixedWidthInt]:
    name = f'int{width}'
    if unsigned:
        name = f'u{name}'

    def _raw(value: int) -> int:
        if not unsigned:
            value = value + 2**(width - 1)
        value = value % (2 ** width)
        if not unsigned:
            value = value - 2**(width - 1)
        return value

    # noinspection PyTypeChecker
    return type(name, (BaseFixedWidthInt,), {
        '_raw': _raw,
        '_width': width,
        '_unsigned': unsigned
    })


int128 = generate_int(128, unsigned=False)
uint128 = generate_int(128, unsigned=True)
int256 = generate_int(256, unsigned=False)
uint256 = generate_int(256, unsigned=True)


# Aliases
sbyte = int8
byte = uint8
short = int16
ushort = uint16
int_ = int32
uint = uint32
long = int64
ulong = uint64


def __getattr__(name: str) -> Type[BaseFixedWidthInt]:
    bits = name
    if bits.startswith('u'):
        unsigned = True
        bits = name[1:]
    else:
        unsigned = False
    if bits.startswith('int'):
        bits = bits[3:]
    else:
        raise AttributeError(name)
    if bits.isnumeric():
        width = int(bits)
    else:
        raise AttributeError(name)
    return generate_int(width, unsigned)


yafwi/__init__.py
"""Yet another fixed width integer (library)."""

from ctypes import (c_int16, c_int32, c_int64, c_int8,
                    c_uint16, c_uint32, c_uint64, c_uint8)
from functools import wraps
from sys import byteorder
from typing import Callable, ClassVar, Optional, Tuple, Type, TypeVar

__all__ = ('FixedWidthInt', 'BaseFixedWidthInt', 'generate_int',
           'int8', 'int16', 'int32', 'int64', 'int128', 'int256',
           'uint8', 'uint16', 'uint32', 'uint64', 'uint128', 'uint256',
           'sbyte', 'byte',
           'short', 'ushort',
           'int_', 'uint',
           'long', 'ulong')
__version__ = '1.0.0'

T = TypeVar('T', bound=int)


class FixedWidthInt(type):
    _raw: Callable[[int], int]
    _width: int
    _unsigned: bool

    @property
    def width(cls) -> int:
        return cls._width

    @property
    def unsigned(cls) -> bool:
        return cls._unsigned

    @property
    def max(cls) -> int:
        if cls._unsigned:
            return cls(2 ** cls._width - 1)
        return cls(2 ** (cls._width - 1) - 1)

    @property
    def min(cls) -> int:
        if cls._unsigned:
            return cls(0)
        return cls(-(2 ** (cls._width - 1)))


def take_wider(fn):
    @wraps(fn)
    def wrapper(self: 'BaseFixedWidthInt', other: int):
        if (isinstance(other, BaseFixedWidthInt)
                and (other.width > self.width
                     or other.width == self.width
                     and other.unsigned
                     and not self.unsigned)):
            return NotImplemented
        return fn(self, other)

    return wrapper


class BaseFixedWidthInt(int, metaclass=FixedWidthInt):
    _raw: Callable[[int], int]
    _width: ClassVar[int]
    _unsigned: ClassVar[bool]

    def __new__(cls, value: int) -> 'BaseFixedWidthInt':
        if cls is BaseFixedWidthInt:
            raise RuntimeError('Use concrete implementation, not base _Int')

        raw = cls._raw(value)
        if hasattr(raw, 'value'):
            return int.__new__(cls, raw.value)
        return int.__new__(cls, raw)

    @property
    def width(self: FixedWidthInt) -> T:
        # noinspection PyTypeChecker
        return type(self).width

    @property
    def unsigned(self: FixedWidthInt) -> bool:
        # noinspection PyTypeChecker
        return type(self).unsigned

    @property
    def max(self: FixedWidthInt) -> T:
        # noinspection PyTypeChecker
        return type(self).max

    @property
    def min(self: FixedWidthInt) -> T:
        # noinspection PyTypeChecker
        return type(self).min

    # Conversion / reprs

    def __bytes__(self) -> bytes:
        return self.to_bytes(self.width // 8,
                             byteorder=byteorder,
                             signed=not self.unsigned)

    @property
    def bin(self) -> str:
        if byteorder == 'little':
            return '0b' + ''.join(f'{b:08b}' for b in reversed(bytes(self)))
        return '0b' + ''.join(f'{b:08b}' for b in bytes(self))

    @property
    def hex(self) -> str:
        if byteorder == 'little':
            return '0x' + ''.join(f'{b:02x}' for b in reversed(bytes(self)))
        return '0x' + ''.join(f'{b:02x}' for b in bytes(self))

    def __repr__(self: T) -> str:
        if self._unsigned:
            prefix = 'u'
        else:
            prefix = ''
        return f'{prefix}int{self.width}({super().__repr__()})'

    # Arithmetic

    @take_wider
    def __add__(self: T, other: int) -> T:
        return type(self)(super().__add__(other))

    @take_wider
    def __radd__(self: T, other: int) -> T:
        return type(self)(super().__radd__(other))

    @take_wider
    def __sub__(self: T, other: int) -> T:
        return type(self)(super().__sub__(other))

    @take_wider
    def __rsub__(self: T, other: int) -> T:
        return type(self)(super().__rsub__(other))

    @take_wider
    def __mul__(self: T, other: int) -> T:
        return type(self)(super().__mul__(other))

    @take_wider
    def __rmul__(self: T, other: int) -> T:
        return type(self)(super().__rmul__(other))

    def __divmod__(self: T, other: int) -> Tuple[T, T]:
        div, mod = super().__divmod__(other)
        return type(self)(div), type(self)(mod)

    def __floordiv__(self: T, other: int) -> T:
        return type(self)(super().__floordiv__(other))

    def __mod__(self: T, other: int) -> T:
        return type(self)(super().__mod__(other))

    def __pow__(self: T, power: int, modulo: Optional[int] = None) -> T:
        return type(self)(super().__pow__(power, modulo))

    # Bitwise ops

    @take_wider
    def __and__(self: T, other: int) -> T:
        return type(self)(super().__and__(other))

    @take_wider
    def __rand__(self: T, other: int) -> T:
        return type(self)(super().__rand__(other))

    @take_wider
    def __or__(self: T, other: int) -> T:
        return type(self)(super().__or__(other))

    @take_wider
    def __ror__(self: T, other: int) -> T:
        return type(self)(super().__ror__(other))

    @take_wider
    def __xor__(self: T, other: int) -> T:
        return type(self)(super().__xor__(other))

    @take_wider
    def __rxor__(self: T, other: int) -> T:
        return type(self)(super().__rxor__(other))

    def __lshift__(self: T, other: int) -> T:
        return type(self)(super().__lshift__(other))

    def __rshift__(self: T, other: int) -> T:
        return type(self)(super().__rshift__(other))

    def __invert__(self: T) -> T:
        return type(self)(super().__invert__())

    # Other maths / conversions

    def __abs__(self: T) -> T:
        return type(self)(super().__abs__())

    def __ceil__(self: T) -> T:
        return type(self)(self)

    def __floor__(self: T) -> T:
        return type(self)(self)

    # pylint: disable=invalid-index-returned
    def __index__(self: T) -> T:
        return type(self)(self)

    def __trunc__(self: T) -> T:
        return type(self)(self)

    def __round__(self: T, n: Optional[int] = None) -> T:
        return type(self)(super().__round__(n))

    def __pos__(self: T) -> T:
        return type(self)(super().__pos__())

    def __neg__(self: T) -> T:
        return type(self)(super().__neg__())


# noinspection PyPep8Naming
class int8(BaseFixedWidthInt):
    _raw = c_int8
    _width = 8
    _unsigned = False


# noinspection PyPep8Naming
class int16(BaseFixedWidthInt):
    _raw = c_int16
    _width = 16
    _unsigned = False


# noinspection PyPep8Naming
class int32(BaseFixedWidthInt):
    _raw = c_int32
    _width = 32
    _unsigned = False


# noinspection PyPep8Naming
class int64(BaseFixedWidthInt):
    _raw = c_int64
    _width = 64
    _unsigned = False


# noinspection PyPep8Naming
class uint8(BaseFixedWidthInt):
    _raw = c_uint8
    _width = 8
    _unsigned = True


# noinspection PyPep8Naming
class uint16(BaseFixedWidthInt):
    _raw = c_uint16
    _width = 16
    _unsigned = True


# noinspection PyPep8Naming
class uint32(BaseFixedWidthInt):
    _raw = c_uint32
    _width = 32
    _unsigned = True


# noinspection PyPep8Naming
class uint64(BaseFixedWidthInt):
    _raw = c_uint64
    _width = 64
    _unsigned = True


def generate_int(width: int, unsigned: bool) -> Type[BaseFixedWidthInt]:
    name = f'int{width}'
    if unsigned:
        name = f'u{name}'

    def _raw(value: int) -> int:
        if not unsigned:
            value = value + 2 ** (width - 1)
        value = value % (2 ** width)
        if not unsigned:
            value = value - 2 ** (width - 1)
        return value

    # noinspection PyTypeChecker
    return type(name, (BaseFixedWidthInt,), {
        '_raw': _raw,
        '_width': width,
        '_unsigned': unsigned
    })


int128 = generate_int(128, unsigned=False)
uint128 = generate_int(128, unsigned=True)
int256 = generate_int(256, unsigned=False)
uint256 = generate_int(256, unsigned=True)

# Aliases
sbyte = int8
byte = uint8
short = int16
ushort = uint16
int_ = int32
uint = uint32
long = int64
ulong = uint64


def __getattr__(name: str) -> Type[BaseFixedWidthInt]:
    bits = name
    if bits.startswith('u'):
        unsigned = True
        bits = name[1:]
    else:
        unsigned = False
    if bits.startswith('int'):
        bits = bits[3:]
    else:
        raise AttributeError(name)
    if bits.isnumeric():
        width = int(bits)
    else:
        raise AttributeError(name)
    return generate_int(width, unsigned)
